<a href="https://colab.research.google.com/github/FabiolaEscobar/Python/blob/main/Miniproyecto1_FabiolaEscobar.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Se importan las librerias necesarias**

In [1]:
import numpy as np
import pandas as pd

# **Se carga y estudia PRIMERA fuente de datos sobre las tarjetas de crédito de los clientes (UCI_Credit_Card_BillATM), contiene 52 registros.**

In [2]:
dfBillATM = pd.read_csv('UCI_Credit_Card_BillATM.csv', sep=',')

In [15]:
dfBillATM.head()

Unnamed: 0  ID LIMIT_BAL     SEX  BILL_AMT1  BILL_AMT2  BILL_AMT3  \
0            0   1   20000.0  female     3913.0     3102.0      689.0   
15          15   1   20000.0  female     3913.0     3102.0      689.0   
1            1   2  120000.0     NaN     2682.0     1725.0     2682.0   
2            2   3   90000.0  female    29239.0    14027.0    13559.0   
3            3   4   50000.0  female    46990.0    48233.0    49291.0   

    BILL_AMT4  BILL_AMT5  BILL_AMT6  
0         0.0        0.0        0.0  
15        0.0        0.0        0.0  
1      3272.0     3455.0     3261.0  
2     14331.0    14948.0    15549.0  
3     28314.0    28959.0    29547.0

In [ ]:
dfBillATM.tail()

In [ ]:
dfBillATM.columns

In [ ]:
dfBillATM.dtypes

In [12]:
dfBillATM

Unnamed: 0  ID LIMIT_BAL     SEX  BILL_AMT1  BILL_AMT2  BILL_AMT3  \
0            0   1   20000.0  female     3913.0     3102.0      689.0   
15          15   1   20000.0  female     3913.0     3102.0      689.0   
1            1   2  120000.0     NaN     2682.0     1725.0     2682.0   
2            2   3   90000.0  female    29239.0    14027.0    13559.0   
3            3   4   50000.0  female    46990.0    48233.0    49291.0   
4            4   5   50000.0    male     8617.0     5670.0    35835.0   
35          35   6   50000.0    male    64400.0    57069.0    57608.0   
5            5   6   50000.0    male    64400.0    57069.0    57608.0   
6            6   7  500000.0    male   367965.0   412023.0   445007.0   
7            7   8  100000.0  female    11876.0      380.0      601.0   
8            8   9  140000.0     NaN    11285.0    14096.0    12108.0   
9            9  10   20000.0    male        0.0        0.0        0.0   
10          10  11  200000.0  female    11073.0     9787.0     5535.0   
11          11  12  260000.0     NaN    12261.0    21670.0     9966.0   
12          12  13  630000.0  female    12137.0     6500.0     6500.0   
13          13  14         -    male    65802.0    67369.0    65701.0   
14          14  15  250000.0    male    70887.0    67060.0    63561.0   
50          50  16   50000.0  female    50614.0    29173.0    28116.0   
16          16  17   20000.0    male    15376.0    18010.0    17428.0   
17          17  18  320000.0    male   253286.0   246536.0   194663.0   
18          18  19         -  female        0.0        0.0        0.0   
19          19  20  180000.0  female        0.0        0.0        0.0   
20          20  21  130000.0  female    38358.0    27688.0    24489.0   
21          21  22  120000.0     NaN      316.0      316.0      316.0   
22          22  23         -  female    41087.0    42445.0    45020.0   
23          23  24  450000.0  female     5512.0    19420.0     1473.0   
24          24  25         -    male     4744.0     7070.0        0.0   
25          25  26   50000.0    male    47620.0    41810.0    36023.0   
26          26  27   60000.0    male     -109.0     -425.0      259.0   
27          27  28   50000.0  female    22541.0    16138.0    17163.0   
28          28  29   50000.0  female      650.0     3415.0     3416.0   
29          29  30   50000.0    male    15329.0    16575.0    17496.0   
30          30  31  230000.0  female    16646.0    17265.0    13266.0   
31          31  32   50000.0     NaN    30518.0    29618.0    22102.0   
32          32  33  100000.0    male    93036.0    84071.0    82880.0   
33          33  34  500000.0  female    10929.0     4152.0    22722.0   
34          34  35  500000.0    male    13709.0     5006.0    31130.0   
51          51  36  160000.0    male    30265.0     -131.0     -527.0   
36          36  37         -    male   186503.0   181328.0   180422.0   
37          37  38   60000.0  female    15054.0     9806.0    11068.0   
38          38  39   50000.0    male        0.0      780.0        0.0   
39          39  40  280000.0    male      498.0     9075.0     4641.0   
40          40  41  360000.0    male   218668.0   221296.0   206895.0   
41          41  42   70000.0  female    67521.0    66999.0    63949.0   
42          42  43   10000.0    male     1877.0     3184.0     6003.0   
43          43  44  140000.0  female    59504.0    61544.0    62925.0   
44          44  45   40000.0  female    18927.0    21295.0    25921.0   
45          45  46  210000.0    male        0.0        0.0        0.0   
46          46  47   20000.0     NaN    14028.0    16484.0    15800.0   
47          47  48  150000.0  female     4463.0     3034.0     1170.0   
48          48  49  380000.0     NaN    22401.0    21540.0    15134.0   
49          49  50         -    male    17447.0    18479.0    19476.0   

    BILL_AMT4  BILL_AMT5  BILL_AMT6  
0         0.0        0.0        0.0  
15        0.0        0.0        0.0  
1      3272.0     34

In [11]:
dfBillATM.count() 

Unnamed: 0    52
ID            52
LIMIT_BAL     52
SEX           45
BILL_AMT1     52
BILL_AMT2     52
BILL_AMT3     52
BILL_AMT4     52
BILL_AMT5     52
BILL_AMT6     52
dtype: int64

Con la función count() identifique que hay 7 registros en esta base que tienen el SEXO vacío

Se ordena el dataframe por ID

In [8]:
dfBillATM.sort_values("ID", inplace=True)

In [6]:
dfBillATM.nunique()

Unnamed: 0    52
ID            50
LIMIT_BAL     27
SEX            2
BILL_AMT1     46
BILL_AMT2     47
BILL_AMT3     45
BILL_AMT4     44
BILL_AMT5     43
BILL_AMT6     43
dtype: int64

Con la función nunique() identifique que hay dos registros duplicados (motivo: total es 52 y el ID tiene 50 valores únicos).

In [17]:
dfBillATM['ID'].duplicated().sum()

2

In [22]:
duplicados = dfBillATM["ID"].duplicated(keep = False)

Con la función duplicate identifico cuales son los registros duplicados en ID y si efectivamente todas sus columnas son iguales

In [23]:
dfBillATM[duplicados]

Unnamed: 0  ID LIMIT_BAL     SEX  BILL_AMT1  BILL_AMT2  BILL_AMT3  \
0            0   1   20000.0  female     3913.0     3102.0      689.0   
15          15   1   20000.0  female     3913.0     3102.0      689.0   
35          35   6   50000.0    male    64400.0    57069.0    57608.0   
5            5   6   50000.0    male    64400.0    57069.0    57608.0   

    BILL_AMT4  BILL_AMT5  BILL_AMT6  
0         0.0        0.0        0.0  
15        0.0        0.0        0.0  
35    19394.0    19619.0    20024.0  
5     19394.0    19619.0    20024.0

Los registros 15 y 35 serán eliminados del dataframe dfBillATM. (Más adelante realizó la eliminación, antes de unificar las fuentes de información)

In [ ]:
pd.isna(dfBillATM)

In [ ]:
pd.isna(dfBillATM).sum()

Con la función isna(df).sum() confirme que para el dato SEXO hay 7 registros con valor vacío. En el estudio de las tres siguientes fuentes de información, si existe el dato SEXO con menos valores null, decido tomar el dato SEXO de esa otra fuente de información

In [ ]:
dfBillATM.describe()

In [ ]:
dfBillATM.describe(include=['object'])

In [ ]:
dfBillATM.describe(include=[np.number])

In [ ]:
dfBillATM.describe(include=[pd.Categorical])

In [ ]:
dfBillATM.describe(include='all')

In [ ]:
dfBillATM.describe(exclude=[np.number])

# **Se carga y estudia SEGUNDA fuente de datos sobre las tarjetas de crédito de los clientes (UCI_Credit_Card_Pay), contiene 52 registros.**

In [24]:
dfPay = pd.read_csv('UCI_Credit_Card_Pay.csv', sep=',')

In [25]:
dfPay.head()

Unnamed: 0  ID LIMIT_BAL       AGE  PAY_0  PAY_2  PAY_3  PAY_4  PAY_5  \
0           0   1   20000.0   No info      2      2     -1     -1     -2   
1           1   2  120000.0  26 years     -1      2      0      0      0   
2           2   3   90000.0  34 years      0      0      0      0      0   
3           3   4   50000.0  37 years      0      0      0      0      0   
4           4   5   50000.0  57 years     -1      0     -1      0      0   

   PAY_6  
0     -2  
1      2  
2      0  
3      0  
4      0

In [7]:
dfPay.tail()

Unnamed: 0  ID LIMIT_BAL       AGE  PAY_0  PAY_2  PAY_3  PAY_4  PAY_5  \
47          47  48  150000.0  46 years      0      0     -1      0      0   
48          48  49  380000.0  32 years     -1     -1     -1     -1     -1   
49          49  50   20000.0  24 years      0      0      0      0      0   
50          50  11         -  34 years      0      0      2      0      0   
51          51  26         -  23 years      0      0      0      0      0   

    PAY_6  
47     -2  
48     -1  
49      0  
50     -1  
51      0

In [8]:
dfPay.columns

Index(['Unnamed: 0', 'ID', 'LIMIT_BAL', 'AGE', 'PAY_0', 'PAY_2', 'PAY_3',
       'PAY_4', 'PAY_5', 'PAY_6'],
      dtype='object')

In [9]:
dfPay.dtypes

Unnamed: 0     int64
ID             int64
LIMIT_BAL     object
AGE           object
PAY_0          int64
PAY_2          int64
PAY_3          int64
PAY_4          int64
PAY_5          int64
PAY_6          int64
dtype: object

In [ ]:
dfPay

In [11]:
dfPay.count() 

Unnamed: 0    52
ID            52
LIMIT_BAL     52
AGE           52
PAY_0         52
PAY_2         52
PAY_3         52
PAY_4         52
PAY_5         52
PAY_6         52
dtype: int64

Con la función count() identifique todos los registros de esta base contienen información en todos sus datos

Se ordena el dataframe por ID

In [12]:
dfPay.sort_values("ID", inplace=True)

In [13]:
dfPay.nunique()

Unnamed: 0    52
ID            50
LIMIT_BAL     24
AGE           26
PAY_0          5
PAY_2          4
PAY_3          4
PAY_4          4
PAY_5          4
PAY_6          4
dtype: int64

Con la función nunique() identifique que hay dos registros duplicados (motivo: total es 52 y el ID tiene 50 valores únicos).

In [26]:
dfPay['ID'].duplicated().sum()

2

In [28]:
duplicadosPay = dfPay["ID"].duplicated(keep = False)

Con la función duplicate identifico cuales son los registros duplicados en ID y si efectivamente todas sus columnas son iguales

In [29]:
dfPay[duplicadosPay]

Unnamed: 0  ID LIMIT_BAL       AGE  PAY_0  PAY_2  PAY_3  PAY_4  PAY_5  \
0            0   1   20000.0   No info      2      2     -1     -1     -2   
10          10   1   20000.0   No info      2      2     -1     -1     -2   
25          25  26         -  23 years      0      0      0      0      0   
51          51  26         -  23 years      0      0      0      0      0   

    PAY_6  
0      -2  
10     -2  
25      0  
51      0

Los registros 10 y 51 serán eliminados del dataframe dfPay. (Más adelante realizó la eliminación, antes de unificar las fuentes de información)

In [14]:
pd.isna(dfPay).sum()

Unnamed: 0    0
ID            0
LIMIT_BAL     0
AGE           0
PAY_0         0
PAY_2         0
PAY_3         0
PAY_4         0
PAY_5         0
PAY_6         0
dtype: int64

Con la función isna(df).sum() confirme que ningún registro contiene datos vacíos

In [15]:
dfPay.describe(include='all')

Unnamed: 0         ID LIMIT_BAL      AGE      PAY_0      PAY_2  \
count    52.000000  52.000000        52       52  52.000000  52.000000   
unique         NaN        NaN        24       26        NaN        NaN   
top            NaN        NaN   50000.0  No info        NaN        NaN   
freq           NaN        NaN         9        8        NaN        NaN   
mean     25.500000  25.038462       NaN      NaN  -0.115385  -0.288462   
std      15.154757  14.687545       NaN      NaN   1.003012   1.054152   
min       0.000000   1.000000       NaN      NaN  -2.000000  -2.000000   
25%      12.750000  12.750000       NaN      NaN  -1.000000  -1.000000   
50%      25.500000  25.500000       NaN      NaN   0.000000   0.000000   
75%      38.250000  37.250000       NaN      NaN   0.000000   0.000000   
max      51.000000  50.000000       NaN      NaN   2.000000   2.000000   

            PAY_3      PAY_4      PAY_5      PAY_6  
count   52.000000  52.000000  52.000000  52.000000  
unique        NaN        NaN        NaN        NaN  
top           NaN        NaN        NaN        NaN  
freq          NaN        NaN        NaN        NaN  
mean    -0.326923  -0.480769  -0.480769  -0.461538  
std      1.115333   0.980004   0.939136   1.110759  
min     -2.000000  -2.000000  -2.000000  -2.000000  
25%     -1.000000  -1.000000  -1.000000  -1.000000  
50%      0.000000   0.000000   0.000000   0.000000  
75%      0.000000   0.000000   0.000000   0.000000  
max      2.000000   2.000000   2.000000   2.000000

In [16]:
dfPay.describe(include=['object'])

LIMIT_BAL      AGE
count         52       52
unique        24       26
top      50000.0  No info
freq           9        8

In [17]:
dfPay.describe(include=[np.number])

Unnamed: 0         ID      PAY_0      PAY_2      PAY_3      PAY_4  \
count   52.000000  52.000000  52.000000  52.000000  52.000000  52.000000   
mean    25.500000  25.038462  -0.115385  -0.288462  -0.326923  -0.480769   
std     15.154757  14.687545   1.003012   1.054152   1.115333   0.980004   
min      0.000000   1.000000  -2.000000  -2.000000  -2.000000  -2.000000   
25%     12.750000  12.750000  -1.000000  -1.000000  -1.000000  -1.000000   
50%     25.500000  25.500000   0.000000   0.000000   0.000000   0.000000   
75%     38.250000  37.250000   0.000000   0.000000   0.000000   0.000000   
max     51.000000  50.000000   2.000000   2.000000   2.000000   2.000000   

           PAY_5      PAY_6  
count  52.000000  52.000000  
mean   -0.480769  -0.461538  
std     0.939136   1.110759  
min    -2.000000  -2.000000  
25%    -1.000000  -1.000000  
50%     0.000000   0.000000  
75%     0.000000   0.000000  
max     2.000000   2.000000

# **Se carga y estudia TERCERA fuente de datos sobre las tarjetas de crédito de los clientes (UCI_Credit_Card_PayATM), contiene 52 registros.**

In [31]:
dfPayATM = pd.read_csv('UCI_Credit_Card_PayATM.csv',sep=',')

In [32]:
dfPayATM.head()

Unnamed: 0  ID LIMIT_BAL EDUCATION  PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  \
0           0   1   20000.0         2       0.0     689.0       0.0       0.0   
1           1   2  120000.0         2       0.0    1000.0    1000.0    1000.0   
2           2   3   90000.0      Null    1518.0    1500.0    1000.0    1000.0   
3           3   4         -         2    2000.0    2019.0    1200.0    1100.0   
4           4   5   50000.0         2    2000.0   36681.0   10000.0    9000.0   

   PAY_AMT5  PAY_AMT6  
0       0.0       0.0  
1       0.0    2000.0  
2    1000.0    5000.0  
3    1069.0    1000.0  
4     689.0     679.0

In [19]:
dfPayATM.tail()

Unnamed: 0  ID LIMIT_BAL EDUCATION  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
47          47  48         -         5    1013.0    1170.0       0.0   
48          48  49  380000.0         2   21540.0   15138.0   24677.0   
49          49  50   20000.0         1    1318.0    1315.0     704.0   
50          50  36  160000.0         1     131.0     396.0     396.0   
51          51  46  210000.0         1       0.0       0.0       0.0   

    PAY_AMT4  PAY_AMT5  PAY_AMT6  
47       0.0       0.0       0.0  
48   11851.0   11875.0    8251.0  
49     928.0     912.0    1069.0  
50     565.0     792.0       0.0  
51       0.0       0.0       0.0

In [20]:
dfPayATM.columns

Index(['Unnamed: 0', 'ID', 'LIMIT_BAL', 'EDUCATION', 'PAY_AMT1', 'PAY_AMT2',
       'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6'],
      dtype='object')

In [21]:
dfPayATM.dtypes

Unnamed: 0      int64
ID              int64
LIMIT_BAL      object
EDUCATION      object
PAY_AMT1      float64
PAY_AMT2      float64
PAY_AMT3      float64
PAY_AMT4      float64
PAY_AMT5      float64
PAY_AMT6      float64
dtype: object

In [22]:
dfPayATM

Unnamed: 0  ID LIMIT_BAL EDUCATION  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
0            0   1   20000.0         2       0.0     689.0       0.0   
1            1   2  120000.0         2       0.0    1000.0    1000.0   
2            2   3   90000.0      Null    1518.0    1500.0    1000.0   
3            3   4         -         2    2000.0    2019.0    1200.0   
4            4   5   50000.0         2    2000.0   36681.0   10000.0   
5            5   6   50000.0         1    2500.0    1815.0     657.0   
6            6   7         -         1   55000.0   40000.0   38000.0   
7            7   8  100000.0         2     380.0     601.0       0.0   
8            8   9  140000.0         3    3329.0       0.0     432.0   
9            9  10   20000.0      Null       0.0       0.0       0.0   
10          10  11  200000.0         3    2306.0      12.0      50.0   
11          11  12  260000.0         1   21818.0    9966.0    8583.0   
12          12  13  630000.0      Null    1000.0    6500.0    6500.0   
13          13  14   70000.0         2    3200.0       0.0    3000.0   
14          14  15  250000.0         1    3000.0    3000.0    3000.0   
15          15  16   50000.0         3       0.0    1500.0    1100.0   
16          16  17   20000.0         1    3200.0       0.0    1500.0   
17          17  18  320000.0         1   10358.0   10000.0   75940.0   
18          18  19  360000.0         1       0.0       0.0       0.0   
19          19  20  180000.0         1       0.0       0.0       0.0   
20          20  21  130000.0         3    3000.0    1537.0    1000.0   
21          21  22  120000.0         2     316.0     316.0       0.0   
22          22  23   70000.0         2    2007.0    3582.0       0.0   
23          23  24  450000.0      Null   19428.0    1473.0     560.0   
24          24  25   90000.0         1    5757.0       0.0    5398.0   
25          25  26   50000.0         3    1973.0    1426.0    1001.0   
26          26  27         -         1       0.0    1000.0       0.0   
27          27  28         -         3    1300.0    1300.0    1000.0   
28          28  29   50000.0         3    3415.0    3421.0    2044.0   
29          29  30   50000.0         1    1500.0    1500.0    1000.0   
30          30  31  230000.0         1   17270.0   13281.0   15339.0   
31          31  32   50000.0         2    1718.0    1500.0    1000.0   
32          32  33  100000.0      Null    3023.0    3511.0    3302.0   
33          33  34  500000.0         2    4152.0   22827.0    7521.0   
34          34  35  500000.0         1    5006.0   31178.0    3180.0   
35          35   1   20000.0         2       0.0     689.0       0.0   
36          36  37  280000.0         2    8026.0    8060.0    6300.0   
37          37  38         -         2    1500.0    1518.0    2043.0   
38          38  39   50000.0         1     780.0       0.0       0.0   
39          39  40  280000.0         1    9075.0       0.0    9976.0   
40          40  41  360000.0         1   10000.0    7000.0    6000.0   
41          41  42   70000.0      Null    3000.0    4500.0    4042.0   
42          42  43   10000.0         2    1500.0    2927.0    1000.0   
43          43  44  140000.0      Null    3000.0    3000.0    3000.0   
44          44  45   40000.0         1    3000.0    5000.0       0.0   
45          45   4         -         2    2000.0    2019.0    1200.0   
46          46  47   20000.0         1    3000.0       0.0   16741.0   
47          47  48         -         5    1013.0    1170.0       0.0   
48          48  49  380000.0         2   21540.0   15138.0   24677.0   
49          49  50   20000.0         1    1318.0    1315.0     704.0   
50          50  36  160000.0         1     131.0     396.0     396.0   
51          51  46  210000.0         1       0.0       0.0       0.0   

    PAY_AMT4  PAY_AMT5  PAY_AMT6  
0        0.0       0.0       0.0  
1     1000.0       0.0    2000.0  
2     1000.0    1000.0    5000.0  
3     1100.0    1069.0    1000.0  
4     9000.0

In [23]:
dfPayATM.count() 

Unnamed: 0    52
ID            52
LIMIT_BAL     52
EDUCATION     52
PAY_AMT1      52
PAY_AMT2      52
PAY_AMT3      52
PAY_AMT4      52
PAY_AMT5      52
PAY_AMT6      52
dtype: int64

Se ordena el dataframe por ID

In [36]:
dfPayATM.sort_values("ID", inplace=True)

In [25]:
dfPayATM.nunique()

Unnamed: 0    52
ID            50
LIMIT_BAL     25
EDUCATION      5
PAY_AMT1      34
PAY_AMT2      35
PAY_AMT3      31
PAY_AMT4      32
PAY_AMT5      33
PAY_AMT6      30
dtype: int64

Con la función nunique() identifique que hay dos registros duplicados (motivo: total es 52 y el ID tiene 50 valores únicos).

In [39]:
dfPayATM['ID'].duplicated().sum()

2

In [40]:
duplicadosPayATM = dfPayATM["ID"].duplicated(keep = False)

Con la función duplicate identifico cuales son los registros duplicados en ID y si efectivamente todas sus columnas son iguales

In [41]:
dfPayATM[duplicadosPayATM]

Unnamed: 0  ID LIMIT_BAL EDUCATION  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
0            0   1   20000.0         2       0.0     689.0       0.0   
35          35   1   20000.0         2       0.0     689.0       0.0   
3            3   4         -         2    2000.0    2019.0    1200.0   
45          45   4         -         2    2000.0    2019.0    1200.0   

    PAY_AMT4  PAY_AMT5  PAY_AMT6  
0        0.0       0.0       0.0  
35       0.0       0.0       0.0  
3     1100.0    1069.0    1000.0  
45    1100.0    1069.0    1000.0

Los registros 35 y 45 serán eliminados del dataframe dfPayATM. (Más adelante realizó la eliminación, antes de unificar las fuentes de información)

In [26]:
pd.isna(dfPayATM).sum()

Unnamed: 0    0
ID            0
LIMIT_BAL     0
EDUCATION     0
PAY_AMT1      0
PAY_AMT2      0
PAY_AMT3      0
PAY_AMT4      0
PAY_AMT5      0
PAY_AMT6      0
dtype: int64

Con la función isna(df).sum() confirme que ningún registro contiene datos vacíos

In [27]:
dfPayATM.describe(include=['object'])

LIMIT_BAL EDUCATION
count         52        52
unique        25         5
top      50000.0         1
freq           8        21

In [28]:
dfPayATM.describe(include=[np.number])

Unnamed: 0         ID      PAY_AMT1      PAY_AMT2      PAY_AMT3  \
count   52.000000  52.000000     52.000000     52.000000     52.000000   
mean    25.500000  24.615385   4814.557692   4920.519231   5199.730769   
std     15.154757  14.973380   8879.389673   8952.014216  12122.670220   
min      0.000000   1.000000      0.000000      0.000000      0.000000   
25%     12.750000  11.750000    945.000000    376.000000     37.500000   
50%     25.500000  24.500000   2003.500000   1500.000000   1000.500000   
75%     38.250000  37.250000   3350.500000   3811.500000   4381.000000   
max     51.000000  50.000000  55000.000000  40000.000000  75940.000000   

            PAY_AMT4       PAY_AMT5      PAY_AMT6  
count      52.000000      52.000000     52.000000  
mean     8909.442308    7749.884615   4025.826923  
std     27900.881776   28067.826843  10730.716777  
min         0.000000       0.000000      0.000000  
25%       325.500000     192.750000      0.000000  
50%      1100.000000    1031.000000    898.500000  
75%      4600.000000    3000.000000   2125.000000  
max    188840.000000  195599.000000  51582.000000

# **Se carga y estudia CUARTA fuente de datos sobre las tarjetas de crédito de los clientes (UCI_Credit_Card_PayNextMonth), contiene 50 registros.**



In [43]:
dfPayNM = pd.read_csv('UCI_Credit_Card_PayNextMonth.csv',sep=',')

In [44]:
dfPayNM.head()

Unnamed: 0  ID SEX  EDUCATION  MARRIAGE   AGE  default.payment.next.month
0           0   1   2        2.0         1  24.0                           1
1           1   2   2        2.0         2  26.0                           1
2           2   3   2        2.0         2  34.0                           0
3           3   4   2        NaN         1  37.0                           0
4           4   5   1        2.0         1  57.0                           0

In [22]:
dfPayNM.tail()

Unnamed: 0  ID  SEX  EDUCATION  MARRIAGE   AGE  default.payment.next.month
45          45  46    1        1.0         2  29.0                           1
46          46  47    2        1.0         2  22.0                           1
47          47  48    2        NaN         2  46.0                           1
48          48  49    1        2.0         2  32.0                           0
49          49  50  Nan        1.0         2  24.0                           0

In [23]:
dfPayNM

Unnamed: 0  ID  SEX  EDUCATION  MARRIAGE   AGE  default.payment.next.month
0            0   1    2        2.0         1  24.0                           1
1            1   2    2        2.0         2  26.0                           1
2            2   3    2        2.0         2  34.0                           0
3            3   4    2        NaN         1  37.0                           0
4            4   5    1        2.0         1  57.0                           0
5            5   6    1        1.0         2  37.0                           0
6            6   7    1        NaN         2  29.0                           0
7            7   8    2        2.0         2  23.0                           0
8            8   9    2        3.0         1  28.0                           0
9            9  10    1        3.0         2  35.0                           0
10          10  11    2        3.0         2   NaN                           0
11          11  12    2        1.0         2  51.0                           0
12          12  13    2        2.0         2  41.0                           0
13          13  14  Nan        2.0         2  30.0                           1
14          14  15    1        1.0         2  29.0                           0
15          15  16    2        3.0         3  23.0                           0
16          16  17    1        1.0         2  24.0                           1
17          17  18    1        1.0         1  49.0                           0
18          18  19  Nan        1.0         1  49.0                           0
19          19  20    2        1.0         2  29.0                           0
20          20  21    2        3.0         2  39.0                           0
21          21  22    2        2.0         1  39.0                           1
22          22  23  Nan        2.0         2  26.0                           1
23          23  24    2        1.0         1  40.0                           1
24          24  25  Nan        1.0         2  23.0                           0
25          25  26    1        3.0         2   NaN                           0
26          26  27    1        NaN         2  27.0                           1
27          27  28    2        NaN         2  30.0                           0
28          28  29    2        3.0         1  47.0                           0
29          29  30    1        1.0         2  26.0                           0
30          30  31    2        1.0         2   NaN                           0
31          31  32    1        2.0         2  33.0                           1
32          32  33    1        1.0         2  32.0                           0
33          33  34    2        2.0         1   NaN                           0
34          34  35    1        1.0         1  58.0                           0
35          35  36    1        1.0         2  30.0                           0
36          36  37  Nan        2.0         1  40.0                           0
37          37  38    2        NaN         2  22.0                           0
38          38  39    1        1.0         2  25.0                           1
39          39  40    1        1.0         2  31.0                           0
40          40  41    1        1.0         2  33.0                           0
41          41  42    2        1.0         2  25.0                           0
42          42  43    1        2.0         2   NaN                           0
43          43  44    2        2.0         1   NaN                           0
44          44  45    2        1.0         2  30.0                           0
45          45  46    1        1.0         2  29.0                           1
46          46  47    2        1.0         2  22.0                           1
47          47  48    2        NaN         2  46.0                           1
48          48  49    1        2.0         2  32.0                           0
49          49  50  Nan        1.0         2  24.0    

In [24]:
dfPayNM.dtypes

Unnamed: 0                      int64
ID                              int64
SEX                            object
EDUCATION                     float64
MARRIAGE                        int64
AGE                           float64
default.payment.next.month      int64
dtype: object

In [45]:
dfPayNM.columns

Index(['Unnamed: 0', 'ID', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE',
       'default.payment.next.month'],
      dtype='object')

In [26]:
dfPayNM.describe(include=['object'])

SEX
count   50
unique   3
top      2
freq    24

In [27]:
dfPayNM.describe(include=[np.number])

Unnamed: 0        ID  EDUCATION   MARRIAGE        AGE  \
count    50.00000  50.00000  44.000000  50.000000  44.000000   
mean     24.50000  25.50000   1.659091   1.760000  33.272727   
std      14.57738  14.57738   0.745317   0.476381   9.519536   
min       0.00000   1.00000   1.000000   1.000000  22.000000   
25%      12.25000  13.25000   1.000000   1.250000  26.000000   
50%      24.50000  25.50000   1.500000   2.000000  30.000000   
75%      36.75000  37.75000   2.000000   2.000000  39.000000   
max      49.00000  50.00000   3.000000   3.000000  58.000000   

       default.payment.next.month  
count                   50.000000  
mean                     0.260000  
std                      0.443087  
min                      0.000000  
25%                      0.000000  
50%                      0.000000  
75%                      0.750000  
max                      1.000000

In [28]:
dfPayNM.count()

Unnamed: 0                    50
ID                            50
SEX                           50
EDUCATION                     44
MARRIAGE                      50
AGE                           44
default.payment.next.month    50
dtype: int64

In [29]:
dfPayNM.nunique()

Unnamed: 0                    50
ID                            50
SEX                            3
EDUCATION                      3
MARRIAGE                       3
AGE                           24
default.payment.next.month     2
dtype: int64

In [46]:
dfPayNM.sort_values("ID", inplace=True)

En esta base, según la función duplicated no hay registros repetidos, de 50 registros hay 50 ids unicos

In [47]:
dfPayNM['ID'].duplicated().sum()

0

In [48]:
pd.isna(dfPayNM).sum()

Unnamed: 0                    0
ID                            0
SEX                           0
EDUCATION                     6
MARRIAGE                      0
AGE                           6
default.payment.next.month    0
dtype: int64

# **Las cuatro fuentes de información serán usadas para crear la base única de información. Los motivos son:**

En la **primera fuente: BillATM.csv** -> se encuentran los datos del BILL_AMT1 al BILL_AMT6 necesarios.

En la **segunda fuente: Pay.csv** -> se encuentran los datos del PAY_0 al PAY_6 necesarios.

En la **tercera fuente: PayATM.csv** -> se encuentran los datos del PAY_AMT1 al PAY_AMT6 necesarios. 

En la **cuarta fuente: PayNextMonth.csv** -> se encuentra el dato default.payment.next.month y el dato MARRIAGE el cual no existe en las otras tres fuentes.

**Sobre el dato SEX** -> existe en la primera y cuarta fuente.

**Sobre el dato LIMIT_BAL** -> existente en la primera, segunda y tercera fuente.

**Sobre el dato AGE**, existente en la segunda y cuarta fuente.

**Sobre el dato EDUCATION**, existente en la tercera y cuarta fuente.

**Sobre los datos MARRIAGE y default payment next month**, los seleccionaré de la cuarta fuente por ser la única donde existen estos datos.

# **Antes de unir las bases, procedo a eliminar los registros duplicados en las tres primeras bases de información y la columna no necesaria 'Unnamed: 0'**

Primera fuente BillATM

In [ ]:
dfBillATM

In [50]:
dfBillATM = dfBillATM.drop([15])

In [51]:
dfBillATM = dfBillATM.drop([35])

In [ ]:
dfBillATM.count()

In [ ]:
dfBillATM

In [53]:
dfBillATM.columns

Index(['Unnamed: 0', 'ID', 'LIMIT_BAL', 'SEX', 'BILL_AMT1', 'BILL_AMT2',
       'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6'],
      dtype='object')

In [54]:
dfBillATM = dfBillATM.drop(['Unnamed: 0'], axis = 1)

Segunda fuente Pay

In [ ]:
dfPay.sort_values("ID", inplace=True)
dfPay

In [57]:
dfPay = dfPay.drop([10])

In [58]:
dfPay = dfPay.drop([51])

In [ ]:
dfPay.nunique()

In [59]:
dfPay.count()

Unnamed: 0    50
ID            50
LIMIT_BAL     50
AGE           50
PAY_0         50
PAY_2         50
PAY_3         50
PAY_4         50
PAY_5         50
PAY_6         50
dtype: int64

In [61]:
dfPay.columns

Index(['Unnamed: 0', 'ID', 'LIMIT_BAL', 'AGE', 'PAY_0', 'PAY_2', 'PAY_3',
       'PAY_4', 'PAY_5', 'PAY_6'],
      dtype='object')

In [62]:
dfPay = dfPay.drop(['Unnamed: 0'], axis = 1)

Tercera fuente PayATM

In [ ]:
dfPayATM.sort_values("ID", inplace=True)
dfPayATM

In [53]:
dfPayATM.columns

Index(['Unnamed: 0', 'ID', 'LIMIT_BAL', 'EDUCATION', 'PAY_AMT1', 'PAY_AMT2',
       'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6'],
      dtype='object')

In [64]:
dfPayATM = dfPayATM.drop([35])

In [65]:
dfPayATM = dfPayATM.drop([45])

In [ ]:
dfPayATM

In [66]:
dfPayATM = dfPayATM.drop(['Unnamed: 0'], axis = 1)

In [70]:
dfPayNM.columns

Index(['ID', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE',
       'default.payment.next.month'],
      dtype='object')

In [69]:
dfPayNM = dfPayNM.drop(['Unnamed: 0'], axis = 1)

# **Continuo con la unión de las cuatro bases de información por el ID único, porque es el dato común en las cuatro fuentes de información**

In [71]:
df = pd.merge(dfBillATM, dfPay, left_on='ID', right_on='ID')

In [72]:
df.columns

Index(['ID', 'LIMIT_BAL_x', 'SEX', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3',
       'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'LIMIT_BAL_y', 'AGE', 'PAY_0',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6'],
      dtype='object')

In [73]:
df = pd.merge(df, dfPayATM, left_on='ID', right_on='ID')

In [74]:
df.columns

Index(['ID', 'LIMIT_BAL_x', 'SEX', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3',
       'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'LIMIT_BAL_y', 'AGE', 'PAY_0',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'LIMIT_BAL', 'EDUCATION',
       'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6'],
      dtype='object')

In [75]:
df = pd.merge(df, dfPayNM, left_on='ID', right_on='ID')

In [76]:
df.columns

Index(['ID', 'LIMIT_BAL_x', 'SEX_x', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3',
       'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'LIMIT_BAL_y', 'AGE_x', 'PAY_0',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'LIMIT_BAL', 'EDUCATION_x',
       'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'SEX_y', 'EDUCATION_y', 'MARRIAGE', 'AGE_y',
       'default.payment.next.month'],
      dtype='object')

# Trabajar sobre dataset de **SEXO**

Tome la decisión de usar la columna SEX_x, porque ya venía con el formato de valores necesatio y además tenia pocos valores NaN, por lo tanto procederé a reemplazarlos por el valor correspondiente de la columna SEX_y

In [77]:
df.iloc[:,2]

0     female
1        NaN
2     female
3     female
4       male
5       male
6       male
7     female
8        NaN
9       male
10    female
11       NaN
12    female
13      male
14      male
15    female
16      male
17      male
18    female
19    female
20    female
21       NaN
22    female
23    female
24      male
25      male
26      male
27    female
28    female
29      male
30    female
31       NaN
32      male
33    female
34      male
35      male
36      male
37    female
38      male
39      male
40      male
41    female
42      male
43    female
44    female
45      male
46       NaN
47    female
48       NaN
49      male
Name: SEX_x, dtype: object

In [78]:
df.iloc[:,-5]

0       2
1       2
2       2
3       2
4       1
5       1
6       1
7       2
8       2
9       1
10      2
11      2
12      2
13    Nan
14      1
15      2
16      1
17      1
18    Nan
19      2
20      2
21      2
22    Nan
23      2
24    Nan
25      1
26      1
27      2
28      2
29      1
30      2
31      1
32      1
33      2
34      1
35      1
36    Nan
37      2
38      1
39      1
40      1
41      2
42      1
43      2
44      2
45      1
46      2
47      2
48      1
49    Nan
Name: SEX_y, dtype: object

In [79]:
df.iloc[:,[2,-5]]

SEX_x SEX_y
0   female     2
1      NaN     2
2   female     2
3   female     2
4     male     1
5     male     1
6     male     1
7   female     2
8      NaN     2
9     male     1
10  female     2
11     NaN     2
12  female     2
13    male   Nan
14    male     1
15  female     2
16    male     1
17    male     1
18  female   Nan
19  female     2
20  female     2
21     NaN     2
22  female   Nan
23  female     2
24    male   Nan
25    male     1
26    male     1
27  female     2
28  female     2
29    male     1
30  female     2
31     NaN     1
32    male     1
33  female     2
34    male     1
35    male     1
36    male   Nan
37  female     2
38    male     1
39    male     1
40    male     1
41  female     2
42    male     1
43  female     2
44  female     2
45    male     1
46     NaN     2
47  female     2
48     NaN     1
49    male   Nan

In [80]:
pd.isna(df.iloc[:,2]).sum()

7

In [81]:
pd.isna(df.iloc[:,-5]).sum()

0

La columna SEX_y no contiene valores NaN, los valores distintos a 1 o 2 son Nan lo cual no es igual a NaN, por ese motivo la función anterior devuelve 0.
Primero, reemplazaré 

In [82]:
df['SEX_y'] = df['SEX_y'].replace('1','female')

In [83]:
df['SEX_y'] = df['SEX_y'].replace('2','male')

In [84]:
df['SEX_y']

0       male
1       male
2       male
3       male
4     female
5     female
6     female
7       male
8       male
9     female
10      male
11      male
12      male
13       Nan
14    female
15      male
16    female
17    female
18       Nan
19      male
20      male
21      male
22       Nan
23      male
24       Nan
25    female
26    female
27      male
28      male
29    female
30      male
31    female
32    female
33      male
34    female
35    female
36       Nan
37      male
38    female
39    female
40    female
41      male
42    female
43      male
44      male
45    female
46      male
47      male
48    female
49       Nan
Name: SEX_y, dtype: object

In [85]:
df['SEX_x']

0     female
1        NaN
2     female
3     female
4       male
5       male
6       male
7     female
8        NaN
9       male
10    female
11       NaN
12    female
13      male
14      male
15    female
16      male
17      male
18    female
19    female
20    female
21       NaN
22    female
23    female
24      male
25      male
26      male
27    female
28    female
29      male
30    female
31       NaN
32      male
33    female
34      male
35      male
36      male
37    female
38      male
39      male
40      male
41    female
42      male
43    female
44    female
45      male
46       NaN
47    female
48       NaN
49      male
Name: SEX_x, dtype: object

Para confirmar si hay NaN en la columna SEX_x

In [86]:
df['SEX_x'].isnull(). values. any()

True

Para confirmar si hay NaN en la columna SEX_y

In [87]:
df['SEX_y'].isnull(). values. any()

False

In [88]:
df['SEX_x'] = df['SEX_x'].fillna(df['SEX_y'])

Con la función anterior **fillna**, logré reemplazar los NaN de la columna SEX_x por el valor que existiese en SEX_y para ese mismo registro. 

Aplicando la función np.where con np.nan vistas en clases no me funciono.

ynan = np.where(df['SEX_x']==np.nan, df['SEX_y'], df['SEX_x'])
df['Sex'] = np.where(df['SEX_x']==df['SEX_y'], df['SEX_x'], ynan)

Incluso, si ejecutaba solamente: df['SEX_x']==np.nan, me tenia que traer TRUE en algunas filas y no lo hace

In [89]:
df = df.drop(labels=["SEX_y"], axis=1)

In [90]:
df.columns

Index(['ID', 'LIMIT_BAL_x', 'SEX_x', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3',
       'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'LIMIT_BAL_y', 'AGE_x', 'PAY_0',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'LIMIT_BAL', 'EDUCATION_x',
       'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'EDUCATION_y', 'MARRIAGE', 'AGE_y', 'default.payment.next.month'],
      dtype='object')

# Trabajar sobre dataset de **EDUCATION**

In [91]:
df.dtypes

ID                              int64
LIMIT_BAL_x                    object
SEX_x                          object
BILL_AMT1                     float64
BILL_AMT2                     float64
BILL_AMT3                     float64
BILL_AMT4                     float64
BILL_AMT5                     float64
BILL_AMT6                     float64
LIMIT_BAL_y                    object
AGE_x                          object
PAY_0                           int64
PAY_2                           int64
PAY_3                           int64
PAY_4                           int64
PAY_5                           int64
PAY_6                           int64
LIMIT_BAL                      object
EDUCATION_x                    object
PAY_AMT1                      float64
PAY_AMT2                      float64
PAY_AMT3                      float64
PAY_AMT4                      float64
PAY_AMT5                      float64
PAY_AMT6                      float64
EDUCATION_y                   float64
MARRIAGE    

In [92]:
df.columns

Index(['ID', 'LIMIT_BAL_x', 'SEX_x', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3',
       'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'LIMIT_BAL_y', 'AGE_x', 'PAY_0',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'LIMIT_BAL', 'EDUCATION_x',
       'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'EDUCATION_y', 'MARRIAGE', 'AGE_y', 'default.payment.next.month'],
      dtype='object')

In [93]:
df.iloc[:,[-11,-4]]

EDUCATION_x  EDUCATION_y
0            2          2.0
1            2          2.0
2         Null          2.0
3            2          NaN
4            2          2.0
5            1          1.0
6            1          NaN
7            2          2.0
8            3          3.0
9         Null          3.0
10           3          3.0
11           1          1.0
12        Null          2.0
13           2          2.0
14           1          1.0
15           3          3.0
16           1          1.0
17           1          1.0
18           1          1.0
19           1          1.0
20           3          3.0
21           2          2.0
22           2          2.0
23        Null          1.0
24           1          1.0
25           3          3.0
26           1          NaN
27           3          NaN
28           3          3.0
29           1          1.0
30           1          1.0
31           2          2.0
32        Null          1.0
33           2          2.0
34           1          1.0
35           1          1.0
36           2          2.0
37           2          NaN
38           1          1.0
39           1          1.0
40           1          1.0
41        Null          1.0
42           2          2.0
43        Null          2.0
44           1          1.0
45           1          1.0
46           1          1.0
47           5          NaN
48           2          2.0
49           1          1.0

Para encontrar y confirmar existencia de valores NaN en ambas columnas de educación

In [94]:
df['EDUCATION_x'].isnull(). values. any()

False

In [95]:
df['EDUCATION_y'].isnull(). values. any()

True

 Tomaré la columna EDUCATION_y por ser de tipo float y tener datos NaN, mientras que la EDUCATION_x es tipo object.

 Primero, reemplazo los valores NaN por 0.
 Segundo, elimino los dos caracteres de la derecha (el punto y el  decimal), dejando unicamente el valor entero.

 Por último, con la función where de libreria Numpy, para los valores 0 de la columna EDUCATION_y, se reemplazará por el valor de la columna EDUCATION_x, de lo contrario debe permanecer el valor de la columna EDUCATION_y

In [96]:
df['EDUCATION_y'] = df['EDUCATION_y'].fillna(0)

In [97]:
df['EDUCATION_y'] = df['EDUCATION_y'].astype(str).str[:-2]

In [98]:
df['EDUCATION_y'] = np.where(df['EDUCATION_y']=='0', df['EDUCATION_x'], df['EDUCATION_y'])

Confirmo que no existen datos NaN en la columna EDUCATION_y

In [99]:
pd.isna(df['EDUCATION_y']).sum()

0

Elimino la columna EDUCATION_x

In [100]:
df = df.drop(labels=["EDUCATION_x"], axis=1)

In [101]:
df_copia=df

Consulto las columnas de mi dataframe

In [102]:
df.columns

Index(['ID', 'LIMIT_BAL_x', 'SEX_x', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3',
       'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'LIMIT_BAL_y', 'AGE_x', 'PAY_0',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'LIMIT_BAL', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'EDUCATION_y', 'MARRIAGE', 'AGE_y', 'default.payment.next.month'],
      dtype='object')

# Trabajar sobre dataset de **MARRIAGE**

Esta columna no se repite, existia en una única fuente de datos

In [103]:
df['MARRIAGE'].dtype

dtype('int64')

In [104]:
df['MARRIAGE'].nunique()

3

In [105]:
pd.isna(df['MARRIAGE']).sum()

0

Se confirma que no existen datos NaN

In [106]:
df['MARRIAGE'].describe()

count    50.000000
mean      1.760000
std       0.476381
min       1.000000
25%       1.250000
50%       2.000000
75%       2.000000
max       3.000000
Name: MARRIAGE, dtype: float64

Con la función describe se evalua que son valores enteros entre 1 y 3. Es decir no hay valores fuera del rango que ocasionen una dispersión anormal.

In [107]:
df['Marriage']=df['MARRIAGE']

In [108]:
df = df.drop(labels=["MARRIAGE"], axis=1)

Se guarda la información en una columna con nombre exacto al solicitado en la fuente única de salida y se elimina la original.

# Trabajar sobre dataset de **AGE**

In [109]:
df.columns

Index(['ID', 'LIMIT_BAL_x', 'SEX_x', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3',
       'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'LIMIT_BAL_y', 'AGE_x', 'PAY_0',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'LIMIT_BAL', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'EDUCATION_y', 'AGE_y', 'default.payment.next.month', 'Marriage'],
      dtype='object')

In [110]:
df.iloc[:,[10,-3]]

AGE_x  AGE_y
0    No info   24.0
1   26 years   26.0
2   34 years   34.0
3   37 years   37.0
4   57 years   57.0
5    No info   37.0
6   29 years   29.0
7   23 years   23.0
8   28 years   28.0
9   35 years   35.0
10  34 years    NaN
11  51 years   51.0
12  41 years   41.0
13  30 years   30.0
14  29 years   29.0
15   No info   23.0
16  24 years   24.0
17  49 years   49.0
18  49 years   49.0
19  29 years   29.0
20   No info   39.0
21  39 years   39.0
22  26 years   26.0
23  40 years   40.0
24  23 years   23.0
25  23 years    NaN
26  27 years   27.0
27  30 years   30.0
28  47 years   47.0
29  26 years   26.0
30  27 years    NaN
31  33 years   33.0
32  32 years   32.0
33  54 years    NaN
34  58 years   58.0
35   No info   30.0
36  40 years   40.0
37  22 years   22.0
38  25 years   25.0
39  31 years   31.0
40   No info   33.0
41  25 years   25.0
42  22 years    NaN
43  37 years    NaN
44  30 years   30.0
45   No info   29.0
46  22 years   22.0
47  46 years   46.0
48  32 years   32.0
49  24 years   24.0

Con describe, se observa que la columna AGE_x es de tipo objeto, conteniendo cadena de caracteres donde la primera es No Info

In [111]:
df['AGE_x'].describe()

count          50
unique         26
top       No info
freq            7
Name: AGE_x, dtype: object

Se confirma que no hay valores Nan

In [112]:
pd.isna(df['AGE_x']).sum()

0

In [113]:
pd.isnull(df['AGE_x']).sum()

0

Con describe, se observa que la columna AGE_y es de tipo float, con una edad minima de 22 y maxima de 58. Un promedio de 33.27. Con estas estadisticas se entiende que los valores de edad son normales, es decir no hay valores como 150 500 1000, que serian numeros de edades no válidos

In [114]:
df['AGE_y'].describe()

count    44.000000
mean     33.272727
std       9.519536
min      22.000000
25%      26.000000
50%      30.000000
75%      39.000000
max      58.000000
Name: AGE_y, dtype: float64

Se confirma que hay 7 valores Nan

In [115]:
pd.isna(df['AGE_y']).sum()

6

In [116]:
df['AGE_y'].head()

0    24.0
1    26.0
2    34.0
3    37.0
4    57.0
Name: AGE_y, dtype: float64

In [117]:
df['AGE_x'].head()

0     No info
1    26 years
2    34 years
3    37 years
4    57 years
Name: AGE_x, dtype: object

Se plantea la condición para identificar los valores No Info de la columna AGE_x

In [118]:
df['AGE_x'] == 'No info'

0      True
1     False
2     False
3     False
4     False
5      True
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15     True
16    False
17    False
18    False
19    False
20     True
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
32    False
33    False
34    False
35     True
36    False
37    False
38    False
39    False
40     True
41    False
42    False
43    False
44    False
45     True
46    False
47    False
48    False
49    False
Name: AGE_x, dtype: bool

Se reemplaza los Nan de la columna AGE_y por 0 y luego se elimina los dos caracteres de la derecha de cada valor flotante

In [119]:
df['AGE_y'] = df['AGE_y'].fillna(0)

In [120]:
df['AGE_y'] = df['AGE_y'].astype(str).str[:-2]

Se plantea la concatenación de la palabra years a la columna AGE_y, de esa forma cuando en la columna AGE_x exista un No info se reemplazará por el valor de la columna AGE_y con el formato deseado

In [121]:
df['AGE_y']+' years'

0     24 years
1     26 years
2     34 years
3     37 years
4     57 years
5     37 years
6     29 years
7     23 years
8     28 years
9     35 years
10     0 years
11    51 years
12    41 years
13    30 years
14    29 years
15    23 years
16    24 years
17    49 years
18    49 years
19    29 years
20    39 years
21    39 years
22    26 years
23    40 years
24    23 years
25     0 years
26    27 years
27    30 years
28    47 years
29    26 years
30     0 years
31    33 years
32    32 years
33     0 years
34    58 years
35    30 years
36    40 years
37    22 years
38    25 years
39    31 years
40    33 years
41    25 years
42     0 years
43     0 years
44    30 years
45    29 years
46    22 years
47    46 years
48    32 years
49    24 years
Name: AGE_y, dtype: object

In [122]:
df['AGE_x'] = np.where(df['AGE_x'] == 'No info', df['AGE_y']+' years', df['AGE_x'])

In [123]:
df.iloc[:,[10,-3]]

AGE_x AGE_y
0   24 years    24
1   26 years    26
2   34 years    34
3   37 years    37
4   57 years    57
5   37 years    37
6   29 years    29
7   23 years    23
8   28 years    28
9   35 years    35
10  34 years     0
11  51 years    51
12  41 years    41
13  30 years    30
14  29 years    29
15  23 years    23
16  24 years    24
17  49 years    49
18  49 years    49
19  29 years    29
20  39 years    39
21  39 years    39
22  26 years    26
23  40 years    40
24  23 years    23
25  23 years     0
26  27 years    27
27  30 years    30
28  47 years    47
29  26 years    26
30  27 years     0
31  33 years    33
32  32 years    32
33  54 years     0
34  58 years    58
35  30 years    30
36  40 years    40
37  22 years    22
38  25 years    25
39  31 years    31
40  33 years    33
41  25 years    25
42  22 years     0
43  37 years     0
44  30 years    30
45  29 years    29
46  22 years    22
47  46 years    46
48  32 years    32
49  24 years    24

In [124]:
df['AGE_x'].nunique()

25

In [125]:
df['AGE_x'].describe()

count           50
unique          25
top       29 years
freq             4
Name: AGE_x, dtype: object

In [126]:
pd.isna(df['AGE_x']).sum()

0

Se confirma que la columna AGE_x almacena la edad en el formato deseado y todos sus campos contienen valor distinto a Nan o null

In [127]:
df = df.drop(labels=["AGE_y"], axis=1)

Se elimina la columna AGE_y

# Trabajar sobre dataset de **default.payment.next.month**

Esta columna solo existía en una de las cuatro fuentes de datos

In [128]:
df['default.payment.next.month'].describe()

count    50.000000
mean      0.260000
std       0.443087
min       0.000000
25%       0.000000
50%       0.000000
75%       0.750000
max       1.000000
Name: default.payment.next.month, dtype: float64

Se detectan valores entre 0 y 1, con un promedio de 0.26.

In [129]:
pd.isna(df['default.payment.next.month']).sum()

0

No se encuentran valores Nan

In [130]:
df['default.payment.next.month'].tail()

45    1
46    1
47    1
48    0
49    0
Name: default.payment.next.month, dtype: int64

In [131]:
df_copia = df

# **Trabajar sobre dataset de LIMIT BAL**

In [132]:
df.columns

Index(['ID', 'LIMIT_BAL_x', 'SEX_x', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3',
       'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'LIMIT_BAL_y', 'AGE_x', 'PAY_0',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'LIMIT_BAL', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'EDUCATION_y', 'default.payment.next.month', 'Marriage'],
      dtype='object')

In [133]:
df.iloc[:,[1,9,17]]

LIMIT_BAL_x LIMIT_BAL_y LIMIT_BAL
0      20000.0     20000.0   20000.0
1     120000.0    120000.0  120000.0
2      90000.0     90000.0   90000.0
3      50000.0     50000.0         -
4      50000.0     50000.0   50000.0
5      50000.0     50000.0   50000.0
6     500000.0    500000.0         -
7     100000.0    100000.0  100000.0
8     140000.0    140000.0  140000.0
9      20000.0     20000.0   20000.0
10    200000.0           -  200000.0
11    260000.0    260000.0  260000.0
12    630000.0    630000.0  630000.0
13           -     70000.0   70000.0
14    250000.0    250000.0  250000.0
15     50000.0     50000.0   50000.0
16     20000.0     20000.0   20000.0
17    320000.0    320000.0  320000.0
18           -    360000.0  360000.0
19    180000.0    180000.0  180000.0
20    130000.0    130000.0  130000.0
21    120000.0    120000.0  120000.0
22           -     70000.0   70000.0
23    450000.0    450000.0  450000.0
24           -     90000.0   90000.0
25     50000.0           -   50000.0
26     60000.0     60000.0         -
27     50000.0     50000.0         -
28     50000.0     50000.0   50000.0
29     50000.0     50000.0   50000.0
30    230000.0           -  230000.0
31     50000.0     50000.0   50000.0
32    100000.0    100000.0  100000.0
33    500000.0           -  500000.0
34    500000.0    500000.0  500000.0
35    160000.0    160000.0  160000.0
36           -    280000.0  280000.0
37     60000.0     60000.0         -
38     50000.0     50000.0   50000.0
39    280000.0    280000.0  280000.0
40    360000.0    360000.0  360000.0
41     70000.0     70000.0   70000.0
42     10000.0           -   10000.0
43    140000.0           -  140000.0
44     40000.0     40000.0   40000.0
45    210000.0    210000.0  210000.0
46     20000.0     20000.0   20000.0
47    150000.0    150000.0         -
48    380000.0    380000.0  380000.0
49           -     20000.0   20000.0

In [134]:
pd.isna(df['LIMIT_BAL_x']).sum()

0

In [135]:
pd.isna(df['LIMIT_BAL']).sum()

0

In [136]:
pd.isna(df['LIMIT_BAL_y']).sum()

0

Se confirma que no existen valores Nan en ninguna de las tres columnas sobre el limite bal

In [137]:
df['LIMIT_BAL_x'].describe()

count          50
unique         27
top       50000.0
freq           10
Name: LIMIT_BAL_x, dtype: object

In [138]:
df['LIMIT_BAL_y'].describe()

count          50
unique         24
top       50000.0
freq            9
Name: LIMIT_BAL_y, dtype: object

In [139]:
df['LIMIT_BAL'].describe()

count          50
unique         25
top       50000.0
freq            8
Name: LIMIT_BAL, dtype: object

Con describe se observa igualdad en el tipo de dato e igual valor superior

In [140]:
(df['LIMIT_BAL_x'] == '-').sum()

6

In [141]:
(df['LIMIT_BAL_y'] == '-').sum()

6

In [142]:
(df['LIMIT_BAL'] == '-').sum()

6

Se detecta que cada columna contiene la misma cantidad de registros con guión '-'

In [143]:
hola = np.where(df['LIMIT_BAL'] == '-',df['LIMIT_BAL_x'],df['LIMIT_BAL'])

In [144]:
hola

array(['20000.0', '120000.0', '90000.0', '50000.0', '50000.0', '50000.0',
       '500000.0', '100000.0', '140000.0', '20000.0', '200000.0',
       '260000.0', '630000.0', '70000.0', '250000.0', '50000.0',
       '20000.0', '320000.0', '360000.0', '180000.0', '130000.0',
       '120000.0', '70000.0', '450000.0', '90000.0', '50000.0', '60000.0',
       '50000.0', '50000.0', '50000.0', '230000.0', '50000.0', '100000.0',
       '500000.0', '500000.0', '160000.0', '280000.0', '60000.0',
       '50000.0', '280000.0', '360000.0', '70000.0', '10000.0',
       '140000.0', '40000.0', '210000.0', '20000.0', '150000.0',
       '380000.0', '20000.0'], dtype=object)

Como la variable temporal hola me guardo la información sin guiones, significa que con aplicar función where de libreria Numpy entre LIMIT_BAL, y LIMIT_BAL_x, es suficiente.

In [145]:
df['LIMIT_BAL'] = np.where(df['LIMIT_BAL'] == '-',df['LIMIT_BAL_x'],df['LIMIT_BAL'])

In [146]:
df.iloc[:,[1,9,17]]

LIMIT_BAL_x LIMIT_BAL_y LIMIT_BAL
0      20000.0     20000.0   20000.0
1     120000.0    120000.0  120000.0
2      90000.0     90000.0   90000.0
3      50000.0     50000.0   50000.0
4      50000.0     50000.0   50000.0
5      50000.0     50000.0   50000.0
6     500000.0    500000.0  500000.0
7     100000.0    100000.0  100000.0
8     140000.0    140000.0  140000.0
9      20000.0     20000.0   20000.0
10    200000.0           -  200000.0
11    260000.0    260000.0  260000.0
12    630000.0    630000.0  630000.0
13           -     70000.0   70000.0
14    250000.0    250000.0  250000.0
15     50000.0     50000.0   50000.0
16     20000.0     20000.0   20000.0
17    320000.0    320000.0  320000.0
18           -    360000.0  360000.0
19    180000.0    180000.0  180000.0
20    130000.0    130000.0  130000.0
21    120000.0    120000.0  120000.0
22           -     70000.0   70000.0
23    450000.0    450000.0  450000.0
24           -     90000.0   90000.0
25     50000.0           -   50000.0
26     60000.0     60000.0   60000.0
27     50000.0     50000.0   50000.0
28     50000.0     50000.0   50000.0
29     50000.0     50000.0   50000.0
30    230000.0           -  230000.0
31     50000.0     50000.0   50000.0
32    100000.0    100000.0  100000.0
33    500000.0           -  500000.0
34    500000.0    500000.0  500000.0
35    160000.0    160000.0  160000.0
36           -    280000.0  280000.0
37     60000.0     60000.0   60000.0
38     50000.0     50000.0   50000.0
39    280000.0    280000.0  280000.0
40    360000.0    360000.0  360000.0
41     70000.0     70000.0   70000.0
42     10000.0           -   10000.0
43    140000.0           -  140000.0
44     40000.0     40000.0   40000.0
45    210000.0    210000.0  210000.0
46     20000.0     20000.0   20000.0
47    150000.0    150000.0  150000.0
48    380000.0    380000.0  380000.0
49           -     20000.0   20000.0

In [147]:
df.columns

Index(['ID', 'LIMIT_BAL_x', 'SEX_x', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3',
       'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'LIMIT_BAL_y', 'AGE_x', 'PAY_0',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'LIMIT_BAL', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'EDUCATION_y', 'default.payment.next.month', 'Marriage'],
      dtype='object')

In [148]:
df = df.drop(labels=["LIMIT_BAL_x","LIMIT_BAL_y"], axis=1)

Comienzo de renombre de columnas y ordenamiento de las columnas

In [149]:
df.columns

Index(['ID', 'SEX_x', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4',
       'BILL_AMT5', 'BILL_AMT6', 'AGE_x', 'PAY_0', 'PAY_2', 'PAY_3', 'PAY_4',
       'PAY_5', 'PAY_6', 'LIMIT_BAL', 'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3',
       'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6', 'EDUCATION_y',
       'default.payment.next.month', 'Marriage'],
      dtype='object')

Genero un respaldo de mi df principal

In [150]:
df_copia =df

In [ ]:
df.columns

# **Generación de archivo .csv con el resultado final**

Con la función reindex, reposicione las columnas incluyendo sus valores

In [152]:
df = df.reindex(columns=['ID', 'LIMIT_BAL', 'SEX_x', 'EDUCATION_y', 'Marriage', 'AGE_x',
                    'PAY_0', 'PAY_2', 'PAY_3', 'PAY_4','PAY_5', 'PAY_6',
                    'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4','BILL_AMT5', 'BILL_AMT6', 
                    'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3','PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',           
                    'default.payment.next.month'])

Genero un respaldo de mi df principal

In [153]:
df_copia =df

Ahora, solo faltaría cambiar el nombre de las columnas 

In [154]:
df.columns = ['ID', 'Limit Balance', 'Sex', 'Education', 'Marriage', 'Age',
                    'PAY 0', 'PAY 2', 'PAY 3', 'PAY 4','PAY 5', 'PAY 6',
                    'BILL AMT1', 'BILL AMT2', 'BILL AMT3', 'BILL AMT4','BILL AMT5', 'BILL AMT6', 
                    'PAY AMT1', 'PAY AMT2', 'PAY AMT3','PAY AMT4', 'PAY AMT5', 'PAY AMT6',           
                   'default payment next month']

In [158]:
df.to_csv('MiniProyecto1_FabiolaEscobar.csv', sep=',')

In [159]:
df_sub2 = pd.read_csv('MiniProyecto1_FabiolaEscobar.csv', sep=',')
df_sub2.head()


Unnamed: 0  ID  Limit Balance     Sex  Education  Marriage       Age  \
0           0   1        20000.0  female          2         1  24 years   
1           1   2       120000.0    male          2         2  26 years   
2           2   3        90000.0  female          2         2  34 years   
3           3   4        50000.0  female          2         1  37 years   
4           4   5        50000.0    male          2         1  57 years   

   PAY 0  PAY 2  PAY 3  ...  BILL AMT4  BILL AMT5  BILL AMT6  PAY AMT1  \
0      2      2     -1  ...        0.0        0.0        0.0       0.0   
1     -1      2      0  ...     3272.0     3455.0     3261.0       0.0   
2      0      0      0  ...    14331.0    14948.0    15549.0    1518.0   
3      0      0      0  ...    28314.0    28959.0    29547.0    2000.0   
4     -1      0     -1  ...    20940.0    19146.0    19131.0    2000.0   

   PAY AMT2  PAY AMT3  PAY AMT4  PAY AMT5  PAY AMT6  \
0     689.0       0.0       0.0       0.0       0.0   
1    1000.0    1000.0    1000.0       0.0    2000.0   
2    1500.0    1000.0    1000.0    1000.0    5000.0   
3    2019.0    1200.0    1100.0    1069.0    1000.0   
4   36681.0   10000.0    9000.0     689.0     679.0   

   default payment next month  
0                           1  
1                           1  
2                           0  
3                           0  
4                           0  

[5 rows x 26 columns]

In [156]:
df.head()

ID Limit Balance     Sex Education  Marriage       Age  PAY 0  PAY 2  \
0   1       20000.0  female         2         1  24 years      2      2   
1   2      120000.0    male         2         2  26 years     -1      2   
2   3       90000.0  female         2         2  34 years      0      0   
3   4       50000.0  female         2         1  37 years      0      0   
4   5       50000.0    male         2         1  57 years     -1      0   

   PAY 3  PAY 4  ...  BILL AMT4  BILL AMT5  BILL AMT6  PAY AMT1  PAY AMT2  \
0     -1     -1  ...        0.0        0.0        0.0       0.0     689.0   
1      0      0  ...     3272.0     3455.0     3261.0       0.0    1000.0   
2      0      0  ...    14331.0    14948.0    15549.0    1518.0    1500.0   
3      0      0  ...    28314.0    28959.0    29547.0    2000.0    2019.0   
4     -1      0  ...    20940.0    19146.0    19131.0    2000.0   36681.0   

   PAY AMT3  PAY AMT4  PAY AMT5  PAY AMT6  default payment next month  
0       0.0       0.0       0.0       0.0                           1  
1    1000.0    1000.0       0.0    2000.0                           1  
2    1000.0    1000.0    1000.0    5000.0                           0  
3    1200.0    1100.0    1069.0    1000.0                           0  
4   10000.0    9000.0     689.0     679.0                           0  

[5 rows x 25 columns]